ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs

```diff
@@ -416,94 +359,134 @@ postconditionImpl model cmdErr resp =
-          -- these blocks, in order to test what happens with duplicated blocks.
-          let testBlock' = testBody {tbIsValid = False }
-          return $ Just $ DuplicateBlock f (TestBlock testHeader' testBlock')
-
--- Many fields of the TestBlock are never used, so we use fixed values.
-genBlock :: Gen TestBlock
-genBlock = do
-    pblockId <- predecessorGenerator
-    isEBB    <- elements [IsNotEBB, IsEBB]
-    slot     <- arbitrary
-    -- We don't care about these
-    let body = TestBody 0 True
-        blockNo = 0
-    return $ mkBlock body (fromOrigin pblockId) slot blockNo isEBB
-
-generatorImpl :: Bool
-              -> Model Symbolic
-              -> Maybe (Gen (At CmdErr Symbolic))
-generatorImpl mkErr m@Model {..} = do
-    genCmd <- generatorCmdImpl m
-    Just $ do
-      At cmd <- genCmd
-      err <- if not (allowErrorFor cmd) then return Nothing
-        else frequency
-          [ (8, return Nothing)
-            -- TODO use the full generator for 'Errors', i.e.
-            -- 'arbitrary'. Now we disable partial writes and
-            -- 'SubstituteWithJunk' corruptions because we cannot
-            -- predict what the model should do with those.
-          , (if mkErr then 1 else 0, Just <$> genErrors False False)]
-      return $ At $ CmdErr cmd err
 
-allowErrorFor :: Cmd -> Bool
-allowErrorFor CreateInvalidFile {} = False
-allowErrorFor CreateFile {}        = False
-allowErrorFor Corrupt {}           = False
-allowErrorFor DuplicateBlock {}    = False
-allowErrorFor _                    = True
+    isEmpty = Map.null blockIdx
+
+    getSlot :: BlockId -> Maybe SlotNo
+    getSlot bid = bslot . fst <$> Map.lookup bid blockIdx
+
+    genSlotStartingFrom :: SlotNo -> Gen SlotNo
+    genSlotStartingFrom slot = chooseSlot slot (slot + 20)
+
+    mbMinMaxSlotInModel :: Maybe (SlotNo, SlotNo)
+    mbMinMaxSlotInModel = do
+      minSlot <- bslot . fst . fst <$> Map.minView blockIdx
+      maxSlot <- bslot . fst . fst <$> Map.maxView blockIdx
+      return (minSlot, maxSlot)
+
+    -- Blocks don't have to be valid, i.e., they don't have to satisfy the
+    -- invariants checked in MockChain and ChainFragment, etc. EBBs don't have
+    -- to have a particular slot number, etc.
+    genTestBlock :: Gen TestBlock
+    genTestBlock = frequency
+      [ (4, genBlockId >>= genSuccessor)
+      , (1, genRandomBlock)
+      ]
 
-blockIdGenerator :: Model Symbolic -> Gen BlockId
-blockIdGenerator Model {..} = do
-    bid <- TestHeaderHash <$> arbitrary
-    elements $ bid : (fst <$> getBlockId dbModel)
+    genSuccessor :: BlockId -> Gen TestBlock
+    genSuccessor prevHash = do
+      body     <- TestBody <$> arbitrary <*> arbitrary
+      slot     <- genSlotStartingFrom $
+        maybe 0 succ (getSlot prevHash)
+      -- We don't care about block numbers in the VolatileDB
+      no       <- BlockNo <$> arbitrary
+      isEBB    <- elements [IsEBB, IsNotEBB]
+      return $ mkBlock body (BlockHash prevHash) slot no isEBB
```

Comment:
Reuse `genRandomBlock`, get out the `body`, `slot`, `no`, `isEBB`, and call `mkBlock` with the right `prevHash`.